src/oData/jsonToOData.ts

```diff
@@ -0,0 +1,175 @@
+	$le: ['maximum', 'formatMaximum'],
+	$gt: ['exclusiveMinimum', 'formatExclusiveMinimum'],
+	$ge: ['minimum', 'formatMinimum'],
+} as const;
+
+// TODO: When using the searchbar, for properties that are enums or oneOf, then we can prefilter (ignore case) the possible values (by labels for oneOf) and only pass those to the final json schema filter object
+// { enum: device.dtmode.oneOf.filter(f => f.title.toLowerCase().includes(search.toLowerCase()).map(f => f.slug) }
+
+// TODO: When ref-scheme or foreign-scheme are used, then we should treat the filter creation as if was a plain direct property and the filter modal should show the appropriate input based on the referenced leaf property schema.
+// eg: if the leaf prop is a number, the modal should show =, >, <.$count
+
+const handlePrimitiveFilter = (
+	key: string,
+	value: JSONSchema & { // ajv extensions
+		regexp?: { pattern?: string; flags?: string };
+		formatMinimum?: string;
+		formatMaximum?: string;
+		formatExclusiveMaximum?: string;
+		formatExclusiveMinimum?: string;
+	},
+): any => {
+	if (value.const != null) {
+		return { [key]: value.const };
+	}
+	if (value.enum != null) {
+		return { [key]: { $in: value.enum } };
+	}
+	const regexp =
+		value.regexp ??
+		(value.pattern != null ? { pattern: value.pattern } : undefined);
+	if (regexp != null) {
+		if (regexp.pattern == null) {
+			throw new Error(
+				`Regex object defined but the pattern property was empty`,
+			);
+		}
+		if (regexp.flags != null && regexp.flags !== 'i') {
+			throw new Error(`Regex flag ${regexp.flags} is not supported`);
+		}
+		if (regexp.flags === 'i') {
+			return {
+				$contains: [{ $tolower: { $: key } }, regexp.pattern.toLowerCase()],
+			};
+		}
+		return { [key]: { $contains: regexp.pattern } };
+	}
+	// Tiny optimization
+	if (typeof value.not !== 'boolean' && value.not?.const != null) {
+		return { [key]: { $ne: value.not.const } };
+	}
+	if (value.not != null && typeof value.not !== 'boolean') {
+		// !properties && !contains && type = string | number | boolean
+		return { $not: handlePrimitiveFilter(key, value.not) };
+	}
+	const filters: any[] = [];
+	for (const [$op, jsonSchemaProps] of Object.entries(comparisonOperatorMap)) {
+		for (const jsonSchemaProp of jsonSchemaProps) {
+			if (value[jsonSchemaProp] != null) {
+				filters.push({ [$op]: value[jsonSchemaProp] });
+			}
+		}
+	}
+	if (filters.length > 0) {
+		return maybePluralOp('$and', filters);
+	}
+	throw new Error('');
+};
+
+const handleFilterArray = (parentKey: string, filterObj: JSONSchema): any => {
+	const filters: any[] = [];
+	const $alias = createAlias(parentKey);
+	if (filterObj.minItems != null && filterObj.minItems > 1) {
+		filters.push({ $ge: [{ [$alias]: { $count: {} } }, filterObj.minItems] });
+	}
+	// Post-MVP: maxItems
+	if (filterObj.contains) {
+		const nestedFilters = filterBuilder([$alias], filterObj.contains);
+		if (nestedFilters) {
+			filters.push(nestedFilters);
+		}
+	}
+
```

Comment:
This needs to work with both nested and top level filters:
```
$expr: {
	$eq: [
		// for now only return releases with one image (service)
		{ ipor: { release_image: { $count: {} } } },
		1,
	],
}
```
```
$filter: {
	$gt: [{ release_image: { $count: {} } }, 2],
},
```

So I think we might need to be doing something like:
```suggestion
const handleFilterArray = (parentKeys: string[], filterObj: JSONSchema): any => {
	const filters: any[] = [];
	const field = parentKeys[parentKeys.length - 1];
	const parentKey = parentKeys.length > 1 ? parentKeys[parentKeys.length - 2] : null
	const $alias = parentKey != null ? createAlias(parentKey) : null;
	if (filterObj.minItems != null && filterObj.minItems > 1) {
		let countObj: AnyObject = { [field]: { $count: {} } };
		if ($alias != null) {
			countObj = { [$alias]: countObj };
		}
		filters.push({ $ge: [countObj, filterObj.minItems] });
	}
	// Post-MVP: maxItems
	if (filterObj.contains) {
		const nestedFilters = filterBuilder([
			...($alias != null ? [$alias] : []),
			field
		], filterObj.contains);
		if (nestedFilters) {
			filters.push(nestedFilters);
		}
	}

```